File: 0189-rotate-array/0189-rotate-array.java
class Solution {
    public void rotate(int[] nums, int k) {
        
        int n = nums.length;
        k = k % n;

        reverseArray(nums, 0, n - k - 1);
        reverseArray(nums, n - k, n - 1);
        reverseArray(nums, 0 , n - 1);
    }

    private void reverseArray(int[] numbers, int startIndex, int endIndex){
            while(startIndex < endIndex) {
                int temp = numbers[startIndex];
                numbers[startIndex] = numbers[endIndex];
                numbers[endIndex] = temp;
                startIndex++;
                endIndex--;
            }
    }
}

File: 0189-rotate-array/0189-rotate-array.kt
class Solution {

    fun rotate(nums: IntArray, k: Int): Unit {
        
        val k = k % nums.size
        swapValue(nums, 0, nums.size- k - 1)
        swapValue(nums, nums.size - k , nums.size - 1)
        swapValue(nums, 0 , nums.size -1)
        return
    }

   fun swapValue(nums: IntArray, start: Int, end: Int) {
    var left = start
    var right = end

    while (left < right) {
        val temp = nums[left]
        nums[left] = nums[right]
        nums[right] = temp

        left++
        right--
    }
   }

}